From 293e8f98e69616aa0298f824c5e3d9b41d4df4cd Mon Sep 17 00:00:00 2001
From: Chen Wandun <chenwandun@huawei.com>
Date: Sat, 12 Nov 2022 16:51:18 +0800
Subject: [PATCH 0032/1398] [Huawei] mm: add config isolation for psi under
 cgroup v1

Offering: HULK
hulk inclusion
category: feature
bugzilla: 187968

-------------------------------

Add CONFIG_PSI_CGROUP_V1 to separate feature of psi under cgroup v1
from baseline.

Signed-off-by: Chen Wandun <chenwandun@huawei.com>
---
 include/linux/psi.h    |  2 ++
 init/Kconfig           | 10 ++++++++++
 kernel/cgroup/cgroup.c |  3 +++
 kernel/sched/cpuacct.c |  2 +-
 kernel/sched/psi.c     | 22 ++++++++++++++--------
 kernel/sched/sched.h   |  2 +-
 6 files changed, 31 insertions(+), 10 deletions(-)

diff --git a/include/linux/psi.h b/include/linux/psi.h
index 2708dd1194ae..ec5ba7954027 100644
--- a/include/linux/psi.h
+++ b/include/linux/psi.h
@@ -14,7 +14,9 @@ struct css_set;
 extern struct static_key_false psi_disabled;
 extern struct psi_group psi_system;
 
+#ifdef CONFIG_PSI_CGROUP_V1
 extern struct static_key_true psi_v1_disabled;
+#endif
 
 void psi_init(void);
 
diff --git a/init/Kconfig b/init/Kconfig
index c4cb83e2ce80..d8fe58b346c4 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -640,6 +640,16 @@ config PSI_DEFAULT_DISABLED
 
 	  Say N if unsure.
 
+config PSI_CGROUP_V1
+       bool "Support PSI under cgroup v1"
+       default n
+       depends on PSI
+       help
+         If set, pressure stall information tracking will be used
+         for cgroup v1 other than v2.
+
+         Say N if unsure.
+
 endmenu # "CPU/Task time and stats accounting"
 
 config CPU_ISOLATION
diff --git a/kernel/cgroup/cgroup.c b/kernel/cgroup/cgroup.c
index d00c391ca19e..501505e6e14a 100644
--- a/kernel/cgroup/cgroup.c
+++ b/kernel/cgroup/cgroup.c
@@ -3721,6 +3721,7 @@ static void cgroup_pressure_release(struct kernfs_open_file *of)
 	psi_trigger_destroy(ctx->psi.trigger);
 }
 
+#ifdef CONFIG_PSI_CGROUP_V1
 struct cftype cgroup_v1_psi_files[] = {
 	{
 		.name = "io.pressure",
@@ -3748,6 +3749,8 @@ struct cftype cgroup_v1_psi_files[] = {
 	},
 	{ } /* terminate */
 };
+#endif
+
 #endif /* CONFIG_PSI */
 
 static int cgroup_freeze_show(struct seq_file *seq, void *v)
diff --git a/kernel/sched/cpuacct.c b/kernel/sched/cpuacct.c
index aa4617a6f8fe..21468ef13b51 100644
--- a/kernel/sched/cpuacct.c
+++ b/kernel/sched/cpuacct.c
@@ -375,7 +375,7 @@ struct cgroup_subsys cpuacct_cgrp_subsys = {
 	.early_init	= true,
 };
 
-#ifdef CONFIG_PSI
+#ifdef CONFIG_PSI_CGROUP_V1
 
 static bool psi_v1_enable;
 static int __init setup_psi_v1(char *str)
diff --git a/kernel/sched/psi.c b/kernel/sched/psi.c
index d87a7d810ffc..d28ca0e53afb 100644
--- a/kernel/sched/psi.c
+++ b/kernel/sched/psi.c
@@ -156,7 +156,10 @@
 static int psi_bug __read_mostly;
 
 DEFINE_STATIC_KEY_FALSE(psi_disabled);
+
+#ifdef CONFIG_PSI_CGROUP_V1
 DEFINE_STATIC_KEY_TRUE(psi_v1_disabled);
+#endif
 
 #ifdef CONFIG_PSI_DEFAULT_DISABLED
 static bool psi_enable;
@@ -763,21 +766,23 @@ static struct psi_group *iterate_groups(struct task_struct *task, void **iter)
 	struct cgroup *cgroup = NULL;
 
 	if (!*iter) {
-		if (static_branch_likely(&psi_v1_disabled))
-			cgroup = task->cgroups->dfl_cgrp;
-		else {
+#ifndef CONFIG_PSI_CGROUP_V1
+		cgroup = task->cgroups->dfl_cgrp;
+#else
 #ifdef CONFIG_CGROUP_CPUACCT
-			if (!cgroup_subsys_on_dfl(cpuacct_cgrp_subsys)) {
+		if (!cgroup_subsys_on_dfl(cpuacct_cgrp_subsys)) {
+			if (!static_branch_likely(&psi_v1_disabled)) {
 				rcu_read_lock();
 				cgroup = task_cgroup(task, cpuacct_cgrp_id);
 				rcu_read_unlock();
-			} else {
-				cgroup = task->cgroups->dfl_cgrp;
 			}
+		} else {
+			cgroup = task->cgroups->dfl_cgrp;
+		}
 #else
-			cgroup = NULL;
+		cgroup = NULL;
+#endif
 #endif
-		}
 	} else if (*iter == &psi_system)
 		return NULL;
 	else
@@ -957,6 +962,7 @@ void psi_memstall_leave(unsigned long *flags)
 		return;
 
 	trace_psi_memstall_leave(_RET_IP_);
+
 	/*
 	 * in_memstall clearing & accounting needs to be atomic wrt
 	 * changes to the task's scheduling state, otherwise we could
diff --git a/kernel/sched/sched.h b/kernel/sched/sched.h
index b2af836ad026..252977131b94 100644
--- a/kernel/sched/sched.h
+++ b/kernel/sched/sched.h
@@ -2671,7 +2671,7 @@ unsigned long scale_irq_capacity(unsigned long util, unsigned long irq, unsigned
 }
 #endif
 
-#ifdef CONFIG_PSI
+#ifdef CONFIG_PSI_CGROUP_V1
 extern struct cftype cgroup_v1_psi_files[];
 #endif
 
-- 
2.17.1

